README.rst
ircpdb - Remotely and collaboratively debug your Python application via an IRC channel
======================================================================================

.. image:: https://travis-ci.org/coddingtonbear/ircpdb.svg?branch=master
    :target: https://travis-ci.org/coddingtonbear/ircpdb

.. image:: https://badge.fury.io/py/ircpdb.svg
    :target: http://badge.fury.io/py/ircpdb

Ircpdb is an adaptation of rpdb that, instead of opening a port and
allowing you to debug over telnet, connects to a configurable IRC
channel so you can collaboratively debug an application remotely.

.. code-block::

    import ircpdb
    ircpdb.set_trace(
        channel="#debugger_hangout",
        limit_access_to=['mynickname'], # List of nicknames that are allowed access
    )

By default, ircpdb will select a nickname on its own and enter the channel
you specify on Freenode, but you can feel free to configure ircpdb to
connect anywhere:

.. code-block::

    import ircpdb
    ircpdb.set_trace(
        channel="#debugger_hangout",
        nickname='im_a_debugger',
        server='irc.mycompany.org',
        limit_access_to=['mynickname', 'someothernickname', 'mybestfriend'],
        port=6667,
        ssl=True,
    )  # See 'Options' below for descriptions of the above arguments

Upon reaching ``set_trace()``, your script will "hang" and the only way to get
it to continue is to access ircpdb by talking to the user that connected to the
above IRC channel.

By default, the debugger will enter the channel you've specified using a
username starting with the hostname of the computer from which it was
launched (in the following example: 'MyHostname').  To interact with
the debugger, just send messages in the channel prefixed with "MyHostname:",
or simply "!".

For example, the following two commands are equivalent, and each will
display the pdb help screen (be sure to replace 'MyHostname' with whatever
username the bot selected)::

    !help

::

    MyHostname: help

Installation
------------

From ``pip``::

    pip install ircpdb

Options
-------

You can either specify the server to connect to using a series of keyword
arguments, or using a single URI string described below in `URI Format`.
If you happen to specify connection parameters using both a URI and
keyword arguments, the keyword arguments will take priority.

* ``uri``: A 'URI' specifying the IRC server and channel to connect to.  If you
  specify a URI, there is no need to specify the below parameters, but if you
  do specify any other parameters, they will override settings specified in the URI.
  See `URI Format` below for more information.
* ``channel`` (**REQUIRED IF NOT USING URI**): The name of the channel (starting with ``#``)
  to connect to on the IRC server.
* ``limit_access_to`` (**REQUIRED IF NOT USING URI**): A list of nicknames that
  are allowed to interact with the debugger.  When specified in a URI, this should
  be a comma-separated list of nicknames.
* ``nickname``: The nickname to use when connecting. Note that an alternate
  name will be selected if this name is already in use. Defaults to using
  the hostname of the machine on which the debugger was executed.
* ``server``: The hostname or IP address of the IRC server.
  Default: ``chat.freenode.net``.
* ``port``: The port number of the IRC server.  Default: ``6697``.
* ``ssl``: Use SSL when connecting to the IRC server?  Default: ``True``.
* ``password``: The server password (if necessary) for the IRC server.
  Default: ``None``.
* ``message_wait_seconds``: The number of seconds that the bot should
  wait between sending messages on IRC.  Many servers, including Freenode,
  will kick clients that send too many messages in too short of a time
  frame.  Default: ``0.8`` seconds.
* ``paste_minimum_response_length``: Try to post messages this length
  or longer to a `gist <http://gist.github.com/>`_ rather than sending
  each line individually via IRC.  This is a useful parameter to use
  if you happen to be connected to a server having very austere
  limits on the number of lines a client can send per minute.
  Default: ``20`` lines.
* ``activation_timeout``: Wait maximally this number of seconds for
  somebody to interact with the debugger in the channel before
  disconnecting and continuing execution.  Default: ``60`` seconds.

Default Settings via Environment Variable
-----------------------------------------

You can specify default connection parameters by setting the ``DEFAULT_IRCPDB_URI``
environment variable with a URI matching the format described below in `URI Format`.

URI Format
----------

Example::

    irc+ssl://botnickname@ircserverhostname:6667/#mychannel?limit_access_to=mynickname

This is a shortcut format to use for specifying IRC connection parameters; roughly
this follows the following format::

    irc[+<ssl?>]://[[<nickname>][:<password>]@]<hostname>[:<port>]/<channel>

All other parameters mentioned in `Options` above can be specified as query string arguments.

Note that this diverges from a standard URI in that you should include the ``#``
characters at the beginning of your channel name **unescaped**.

Use in Django Templates
-----------------------

In your `settings.py`, add `ircpbd.django` to your installed apps::

    INSTALLED_APPS = [
        # Other apps
        # ...
        'ircpdb.django',
    ]

Within the template you'd like to add a debugger trace to, load the
`ircpdb` template tags by adding the following to the top of the template::

    {% load ircpdb %}

And, where you'd like to inject the ircpdb trace::

    {% set_trace channel='#my_channel' limit_access_to='coddingtonbear' %}

.. note::

   Although most parameters are unchanged between when invoking ``set_trace``
   in python and invoking ``set_trace`` from within a template, the parameter
   ``limit_access_to`` should be a comma-separated list of usernames rather
   than a list literal when using ``set_trace`` in a template (like above).

Next time you render this template (probably by going to a view that
uses it), rendering will be halted at the point where you've placed your trace,
and the ircpdb bot will appear in your channel.

Security Disclaimer
-------------------

The way that this library works is **inherently** **dangerous**; given that
you're able to execute arbitrary Python code from within your debugger,
it is strongly recommended that you take all reasonable measures to ensure
that you control who are able to execute debugger commands.

To limit your risk as much as possible, you should consider taking the
following steps:

* Always use an SSL-capable IRC server (read: leave the ``ssl`` argument
  set to it's default: ``True``).
* Connect to an IRC server you or a company you work for owns rather than
  Freenode (the default).

Just to make absolutely sure this is clear: you're both responsible for
determining what level of risk you are comfortable with, and for taking
appropriate actions to mitigate that risk.

As is clearly and thunderously stated library's license (see the included
``LICENSE.txt``)::

    THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
    ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
    ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
    FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
    DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
    OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
    HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
    OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
    SUCH DAMAGE.

Good luck, and happy debugging!

Troubleshooting
---------------

If you do not see the bot entering your specified channel, try increasing
the logging level by adding the following lines above your trace to gather
a little more information about problems that may have occurred while 
connecting to the IRC server:

.. code-block::

   import logging
   logging.basicConfig(filename='/path/to/somewhere.log', level=logging.DEBUG)

Author(s)
---------
Adam Coddington <me@adamcoddington.net> - http://adamcoddington.net/

This library is a fork of rpdb, and the underpinnings of this library
are owed to Bertrand Janin <b@janin.com> - http://tamentis.com/ and
all other contributors to `rpdb <https://github.com/tamentis/rpdb>`
including the following:

 - Ken Manheimer - @kenmanheimer
 - Steven Willis - @onlynone
 - Jorge Niedbalski R <niedbalski@gmail.com>
 - Cyprien Le Pannérer <clepannerer@edd.fr>
 - k4ml <kamal.mustafa@gmail.com>
 - Sean M. Collins <sean@coreitpro.com>


CHANGES.txt
1.8 (2016-08-19)
================

 - Switched to Gist instead of Dpaste.
 - Added functionality allowing one to specify their own Paste backend.

1.7 (2015-01-28)
================

 - Added 'activation_timeout' timer that defines a time limit for the first
   interaction with the IRC channel before disconnecting and continuing
   execution.

1.6 (2014-12-20)
================

 - Adds a django template tag {% set_trace %}; this is usable for opening
   up an ircpdb channel during template parsing.

1.5 (2014-12-04)
================

 - Adds functionality for specifying configuration parameters via a single
   argument containing a URI.  Rather than specifying potentially a dozen
   separate parameters, one can now provide a single string; for example:
   "irc+ssl://bot@chat.freenode.net:6667/#mychannel?limit_access_to=me".
 - Adds ability to specify default configuration parameters via environment
   variable named 'DEFAULT_IRCPDB_URI'.

1.4 (2014-11-11)
================

 - Simpler internal handling of prompt.
 - Adding internal support for IRC CTCP commands.
 - Pinning jaraco/irc library version at 0.9 to preserve Python 2.6 support.

1.3 (2014-11-01)
================

 - Reduced security concerns by making it impossible to configure the
   debugger to respond to all usernames.
 - Fixed a UI glitch in which the prompt would be displayed in inappropriate
   situations.

1.2 (2014-10-31)
================

 - Many UI improvements including:
   - Display a prompt after the response has been written.
 - Added '!!forbid' command.
 - Added a thunderous warning to the readme about the security implications
   of this library.

1.1 (2014-10-30)
================

 - Added functionality for automatically sending long messages to Dpaste.
 - Added additional commands including `!!set_dpaste_minimum_response_length`
   and `!!set_message_wait_seconds`.
 - Send a hello message once connected to the channel.

1.0 (2014-10-29)
==================

 - Forked from rpdb to, instead of opening a socket to interact with,
   opening a connection to an IRC channel.
   
0.1.5 (2014-10-16)
==================

 - Write addr/port to stderr instead of stdout (thanks to @onlynone).
 - Allow for dynamic host port (thanks to @onlynone).
 - Make q/quit do proper cleanup (@kenmanheimer)
 - Benignly disregard repeated rpdb.set_trace() to same port as currently
   active session (@kenmanheimer)
 - Extend backwards compatibility down to Python 2.5 (@kenmanheimer)

0.1.4 (2014-04-28)
==================

 - Expose the addr, port arguments to the set_trace method (thanks to @niedbalski).

0.1.3 (2013-08-02)
==================

 - Remove a try/finally that seemed to shift the trace location (thanks to k4ml@github).

0.1.2 (2012-01-26)
==================

 - Catch IOError raised by print in initialization, it may not work in some
   environments (e.g. mod_wsgi). (Menno Smits)

0.1.1 (2010-05-09)
==================

Initial release.



ircpdb/debugger.py
import logging
import os
import pdb
import socket
import ssl as ssllib
import sys
from threading import Thread
import traceback

from irc.connection import Factory
import six

from . import paste_backends
from .bot import IrcpdbBot
from .exceptions import NoAllowedNicknamesSelected, NoChannelSelected
from .parse import parse_irc_uri


logger = logging.getLogger(__name__)


DEFAULT_PARAMS = {
    'channel': None,
    'nickname': None,
    'server': 'chat.freenode.net',
    'port': 6697,
    'password': None,
    'ssl': True,
    'limit_access_to': None,
    'message_wait_seconds': 0.8,
    'paste_minimum_response_length': 20,
    'activation_timeout': 60
}


class Ircpdb(pdb.Pdb):
    def __init__(self, uri=None, **kwargs):
        """Initialize the socket and initialize pdb."""
        params = DEFAULT_PARAMS.copy()
        params.update(parse_irc_uri(uri))
        params.update(kwargs)

        # Backup stdin and stdout before replacing them by the socket handle
        self.old_stdout = sys.stdout
        self.old_stdin = sys.stdin
        self.read_timeout = 0.1

        if not params.get('limit_access_to'):
            raise NoAllowedNicknamesSelected(
                "You must specify a list of nicknames that are allowed "
                "to interact with the debugger using the "
                "`limit_access_to` keyword argument."
            )
        elif isinstance(params.get('limit_access_to'), six.string_types):
            params['limit_access_to'] = [params.get('limit_access_to')]

        connect_params = {}
        if not params.get('nickname'):
            params['nickname'] = socket.gethostname().split('.')[0]
        if not params.get('channel'):
            raise NoChannelSelected(
                "You must specify a channel to connect to using the "
                "`channel` keyword argument."
            )
        if params.get('ssl'):
            connect_params['connect_factory'] = (
                Factory(wrapper=ssllib.wrap_socket)
            )

        # Writes to stdout are forbidden in mod_wsgi environments
        try:
            logger.info(
                "ircpdb has connected to %s:%s on %s\n",
                params.get('server'),
                params.get('port'),
                params.get('channel')
            )
        except IOError:
            pass

        r_pipe, w_pipe = os.pipe()
        # The A pipe is from the bot to pdb
        self.p_A_pipe = os.fdopen(r_pipe, 'r')
        self.b_A_pipe = os.fdopen(w_pipe, 'w')

        r_pipe, w_pipe = os.pipe()
        # The B pipe is from pdb to the bot
        self.b_B_pipe = os.fdopen(r_pipe, 'r')
        self.p_B_pipe = os.fdopen(w_pipe, 'w')

        pdb.Pdb.__init__(
            self,
            stdin=self.p_A_pipe,
            stdout=self.p_B_pipe,
        )

        paste_backend = paste_backends.GistBackend()
        self.bot = IrcpdbBot(
            channel=params.get('channel'),
            nickname=params.get('nickname'),
            server=params.get('server'),
            port=params.get('port'),
            password=params.get('password'),
            limit_access_to=params.get('limit_access_to'),
            message_wait_seconds=params.get('message_wait_seconds'),
            paste_minimum_response_length=(
                params.get('paste_minimum_response_length')
            ),
            paste_backend=paste_backend,
            activation_timeout=params.get('activation_timeout'),
            **connect_params
        )

    def shutdown(self):
        """Revert stdin and stdout, close the socket."""
        sys.stdout = self.old_stdout
        sys.stdin = self.old_stdin
        pipes = [
            self.p_A_pipe,
            self.p_B_pipe,
            self.b_A_pipe,
            self.b_B_pipe
        ]
        for pipe in pipes:
            try:
                pipe.close()
            except IOError:
                logger.warning(
                    "IOError encountered while closing a pipe; messages "
                    "may have been lost."
                )
        self.bot.disconnect()

    def do_continue(self, arg):
        """Clean-up and do underlying continue."""
        try:
            return pdb.Pdb.do_continue(self, arg)
        finally:
            self.shutdown()

    do_c = do_cont = do_continue

    def do_quit(self, arg):
        """Clean-up and do underlying quit."""
        try:
            return pdb.Pdb.do_quit(self, arg)
        finally:
            self.shutdown()

    do_q = do_exit = do_quit


def set_trace(*args, **kwargs):
    """Wrapper function to keep the same import x; x.set_trace() interface.

    We catch all the possible exceptions from pdb and cleanup.

    """
    if not args and 'DEFAULT_IRCPDB_URI' in os.environ:
        args = (
            os.environ['DEFAULT_IRCPDB_URI'],
        )
    debugger = Ircpdb(*args, **kwargs)
    try:
        irc_feeder = Thread(
            target=debugger.bot.process_forever,
            args=(debugger.b_B_pipe, debugger.b_A_pipe, ),
        )
        irc_feeder.daemon = True
        irc_feeder.start()

        debugger.set_trace(sys._getframe().f_back)
    except Exception:
        debugger.shutdown()
        traceback.print_exc()
